driver-core/src/main/java/com/datastax/driver/core/Connection.java

```diff
@@ -492,23 +492,51 @@ void setKeyspace(String keyspace) throws ConnectionException {
             return thisFuture;
         }
 
-        logger.trace("{} Setting keyspace {}", this, keyspace);
-        // Note: we quote the keyspace below, because the name is the one coming from Cassandra, so it's in the right case already
-        Future future = write(new Requests.Query("USE \"" + keyspace + '"'));
-        return Futures.transform(future, new Function<Message.Response, Connection>() {
-            @Override
-            public Connection apply(Message.Response response) {
-                if (response instanceof SetKeyspace) {
-                    Connection.this.keyspace = ((SetKeyspace) response).keyspace;
-                    return Connection.this;
-                } else if (response.type == ERROR) {
-                    Responses.Error error = (Responses.Error) response;
-                    throw defunct(error.asException(address));
-                } else {
-                    throw defunct(new DriverInternalError("Unexpected response while setting keyspace: " + response));
+        SetKeyspaceAttempt attempt = new SetKeyspaceAttempt(keyspace);
+
+        // Check for an existing keyspace attempt.
+        while (true) {
+            SetKeyspaceAttempt existingAttempt = targetKeyspace.get();
+            // if existing attempts' keyspace matches what we are trying to set, use it.
+            if (attempt.equals(existingAttempt)) {
+                attempt = existingAttempt;
+                break;
+            } else if (targetKeyspace.compareAndSet(existingAttempt, attempt)) {
```

Comment:
If the CAS fails, we want to wait on the new attempt stored in the atomic reference.
This will likely be a setKeyspace to the same keyspace that we raced with. Note that it could also target a different keyspace if the client program is constantly switching keyspaces, but in that case there's not much we can do to prevent requests from failing.